ledger/notary-plugin/src/main/java/net/corda/v5/ledger/notary/plugin/core/NotaryException.java

```diff
@@ -0,0 +1,58 @@
+package net.corda.v5.ledger.notary.plugin.core;
+
+import net.corda.v5.base.annotations.CordaSerializable;
+import net.corda.v5.base.exceptions.CordaRuntimeException;
+import net.corda.v5.crypto.SecureHash;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Representation of errors that can be returned by the notary (plugin). This is only a marker interface, the plugins
+ * can define their own errors by implementing this interface. Please refer to the non-validating notary plugin for
+ * a more detailed example.
+ */
+@CordaSerializable
+public abstract class NotaryException extends CordaRuntimeException {
+
+    @NotNull
+    private final String notaryErrorMessage;
+    @Nullable
+    private final SecureHash txId;
+
+    /**
+     * @return the specific error message produced by the notary.
+     */
+    @NotNull
+    public final String getNotaryErrorMessage() {
+        return this.notaryErrorMessage;
+    }
+
+    /**
+     * @return txId Id of the transaction to be notarised. Can be _null_ if an error occurred before the id could be
+     * resolved.
+     */
+    @Nullable
+    public final SecureHash getTxId() {
+        return this.txId;
+    }
+
+
+    private NotaryException(@NotNull String notaryErrorMessage, @Nullable SecureHash txId) {
+        super("Unable to notarise transaction " + (txId != null ? txId : "<Unknown>:") + " " + notaryErrorMessage);
+        this.notaryErrorMessage = notaryErrorMessage;
+        this.txId = txId;
+    }
+
+    @CordaSerializable
+    public abstract static class NotaryExceptionFatal extends NotaryException {
```

Comment:
This and the unknown errors feel a bit clunky; callers now have to qualify the exception, e.g. `NotaryException.NotaryExceptionFatal`. If the intent is to replicate Kotlin sealed class functionality to prevent direct inheritance from `NotaryException`, you might want to consider making `NotaryException` "package private", i.e. specifying no qualifer. This will have a similar effect, in that whilst it can be extended, it could only be extended by R3 (only we should be using the `net.corda.v5.ledger.notary.plugin.core` package, and presumably sandboxing enforces this).

This would also allow you to define the fatal / unknown classes in their own files in this package and make things a bit cleaner.

If we do keep as-is, consider renaming to `Fatal` and `Unknown`, so at least then to the caller it appears as `NotaryException.Fatal`, which avoids duplication.

Reply:
I do agree with it being clunky like this and happy to modify it to `Fatal` and `Unknown`. I'm not a big fan of using package private but if you feel like that would make it cleaner I can make the change.

Reply:
Maybe worth getting others opinions too - people like @lankydan and @MrMatthewLayton are probably better judges than me.

Reply:
I made the change but can make it back if the others feel like it is better. You have a point that static inner classes are not really java style and can look weird if making a static import.

Like let's say someone statically imports the `NotaryException` and then they only write `Fatal` and `Unknown` without the `NotaryException.` prefix that can lead to major confusion too.